docs/source/extending/overloading-guide.rst

```diff
@@ -0,0 +1,194 @@
+
+.. _overloading-guide:
+
+==============================
+A guide to using ``@overload``
+==============================
+
+
+As mentioned in the :ref:`high-level extension API <high-level-extending>`, you
+can use the ``@overload`` decorator to create a Numba implementation of a
+function that can be used in :term:`nopython mode` functions. A common use case
+is to re-implement NumPy functions so that they can be called in ``@jit``
+decorated code. This section discusses how and when to use the
+``@overload`` decorator and what contributing such a function to the Numba code base might
+entail. This should help you get started when needing to use the ``@overload``
+decorator or when attempting to contribute new functions to Numba itself.
+
+The ``@overload`` decorator and it's variants are useful when you have a
+third-party library that you do not control and you wish to provide Numba
+compatible implementations for specific functions from that library.
+
+Concrete Example
+================
+
+Let's assume that you are working on a minimization algorithm that makes use of
+|scipy.linalg.norm|_ to find different vector norms and the `frobenius
+norm <https://en.wikipedia.org/wiki/Frobenius_inner_product>`_ for matrices.
+You know that only integer and real numbers will be involved. (While this may
+sound like an artificial example, especially because a Numba implementation of
+``numpy.linalg.norm`` exists, it is largely pedagogical and serves to
+illustrate how and when to use ``@overload``).
+
+.. |scipy.linalg.norm| replace:: ``scipy.linalg.norm``
+.. _scipy.linalg.norm: https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.norm.html
+
+The skeleton might look something like this::
+
+    def algorithm():
+        # setup
+        v = ...
+        while True:
+            # take a step
+            d = scipy.linalg.norm(v)
+            if d < tolerance:
+                break
+
+Now, let's further assume, that you have heard of Numba and you now wish to use
+it to accelerate your function. However, after adding the
+``jit(nopython=True)``
+decorator, Numba complains that ``scipy.linalg.norm`` isn't supported. From
+looking at the documentation, you realize that a norm is probably fairly easy
+to implement using NumPy and you wanted to learn a little about the inner
+workings of Numba anyway. So, after some back and forth with the
+core-developers via Gitter you receive the following annotated template in your
+INBOX:
+
+.. literalinclude:: template.py
+
+After some deliberation and tinkering, you end up with the following code:
+
+.. literalinclude:: mynorm.py
+
+As you can see, the implementation only supports what you need right now:
+
+* Only supports 
```

Comment:
does this sentence have an end? if not perhaps remove it?!

Reply:
It should be part of the next bullet.